src/Features/Core/Portable/IntroduceUsingStatement/AbstractIntroduceUsingStatementCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,163 @@
+﻿using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.IntroduceUsingStatement
+{
+    internal abstract class AbstractIntroduceUsingStatementCodeRefactoringProvider<TLocalDeclarationSyntax> : CodeRefactoringProvider
+        where TLocalDeclarationSyntax : SyntaxNode
+    {
+        protected abstract Task<Document> IntroduceUsingStatementAsync(
+            Document document,
+            TLocalDeclarationSyntax declarationSyntax,
+            ILocalSymbol localVariable,
+            CancellationToken cancellationToken);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var (declarationSyntax, _) =
+                await FindDisposableLocalDeclaration(
+                    context.Document,
+                    context.Span,
+                    context.CancellationToken).ConfigureAwait(false);
+
+            if (declarationSyntax != null)
+            {
+                context.RegisterRefactoring(CodeAction.Create(
+                    "Introduce using statement",
+                    async cancellationToken =>
```

Comment:
Small nit: in general, we prefer the 'refactoring action' code to be in a sibling function.  But i don't feel super strong about it.

Reply:
Call the higher level method IntroduceUsingStatementAsync  since that's what it is doing.  Call the helpers it calls more appropriate names :)